tools/python/test_custom_ops_export.py

```diff
@@ -0,0 +1,154 @@
+                 batch_size=2, state_dict=None,
+                 rtol=0.001, atol=1e-7,
+                 do_constant_folding=True,
+                 dynamic_axes=None, test_with_inputs=None,
+                 input_names=None, output_names=None,
+                 fixed_batch_size=False):
+        model.eval()
+
+        if input is None:
+            input = torch.randn(batch_size, 3, 224, 224, requires_grad=True)
+
+        with torch.no_grad():
+            if isinstance(input, torch.Tensor):
+                input = (input,)
+            # In-place operators will update input tensor data as well.
+            # Thus inputs are replicated before every forward call.
+            input_copy = copy.deepcopy(input)
+            output = model(*input_copy)
+            if isinstance(output, torch.Tensor):
+                output = (output,)
+
+            # export the model to ONNX
+            f = io.BytesIO()
+            input_copy = copy.deepcopy(input)
+            torch.onnx._export(model, input_copy, f,
+                               opset_version=self.opset_version,
+                               example_outputs=output,
+                               do_constant_folding=do_constant_folding,
+                               keep_initializers_as_inputs=self.keep_initializers_as_inputs,
+                               dynamic_axes=dynamic_axes,
+                               input_names=input_names, output_names=output_names,
+                               fixed_batch_size=fixed_batch_size,
+                               custom_opsets=custom_opsets)
+
+            # compute onnxruntime output prediction
+            ort_sess = onnxruntime.InferenceSession(f.getvalue())
+            input_copy = copy.deepcopy(input)
+            ort_test_with_input(ort_sess, input_copy, output, rtol, atol)
+
+            # if additional test inputs are provided run the onnx
+            # model with these inputs and check the outputs
+            if test_with_inputs is not None:
+                for test_input in test_with_inputs:
+                    if isinstance(test_input, torch.Tensor):
+                        test_input = (test_input,)
+                    test_input_copy = copy.deepcopy(test_input)
+                    output = model(*test_input_copy)
+                    if isinstance(output, torch.Tensor):
+                        output = (output,)
+                    ort_test_with_input(ort_sess, test_input, output, rtol, atol)
+
+    def test_inverse(self):
+        class CustomInverse(torch.nn.Module):
+            def forward(self, x):
+                return torch.inverse(x) + x
+
+        x = torch.randn(2, 3, 3)
+        self.run_test(CustomInverse(), x, custom_opsets={'com.microsoft': 1})
+
+    def test_gelu(self):
+        model = torch.nn.GELU()
+        x = torch.randn(3, 3)
+        self.run_test(model, x, custom_opsets={'com.microsoft': 1})
+
+
+# opset 10 tests
+TestONNXRuntime_opset10 = type(str("TestONNXRuntime_opset10"),
```

Comment:
Are all the new types below being used somewhere? 
Could we use a loop to prevent code duplication?

Reply:
This is a good point. We can even discuss if all these for different opsets are required or not.
Since the purpose here is to ONNXRuntime custom ops, and those ops are registered as part of a custom domain, the ONNX models in these tests will have a combination of standard ONNX ops + Custom ops (from custom microsoft domain). 
For example, a model can have ```com.microsoft::Inverse``` op followed by ```onnx::add```.
And in this case, opset version for com.microsoft will be different than ONNX opset version.

(I've noticed that ONNX Runtime ops are only registered for a single opset version `1`. )
I don't think it's adding much value to test the model export here for different ONNX opset versions. Maybe just testing default opset version would be enough.

Reply:
My understanding was that line 110 merely creates a new type and didn't run anythings as it only contains `new_var = type(name, bases, dict)`

Reply:
It actually creates new test cases since these are subclasses of unittest.TestCase.
So overall we have 16 (8 x 2) tests in this file now.